[fonts] ability to capitalize the first letter of each word

mkortstiege opened the pull request:
This adds the ability to capitalize the first letter of each word.
Usage: Same as upper-/lowercase by using `[CAPIZALIZE]` or `capitalize` in the font.xml style tag.

Request at forums: http://forum.kodi.tv/showthread.php?tid=220249

tamland commented:
I think this style is normally called sentence case.

mkortstiege commented:
Can't follow. You're just commenting on the wording in the commit and PR message or do you want me to change something?

tamland commented:
Yeah, call the tag `[SENTENCECASE]` and `ToSentence` or `ToSentenceCase` etc. so it's in line with the other (upper and lower) letter casing.

tamland commented on xbmc/utils/StringUtils.cpp:
```diff
@@ -354,6 +355,18 @@ wchar_t toupperUnicode(const wchar_t& c)
+
+   Type operator()(Type c)
+   {
+      return std::isspace(c) ? (bFirst = true, c) : bFirst ? (bFirst = false, std::toupper(c)) : c;
```
I don't understand any of that:) Can you make it more verbose so it's clearer that is changes state and what else doing on?

Paxxi replied:
this doesn't compile, std::isspace and std::toupper takes two parameters, you probably meant ::isspace and ::toupper to access the c version of them

mkortstiege commented:
I am actually reading sentence casing the first time today. For me this is known as capitalizing .. Would like to hear other opinions before changing.

mkortstiege commented:
An alternative to the templating would be something like this.
```
vector<string> words = Split(str, " ");
for (unsigned int i = 0; i < words.size(); i++)
  words[i][0] = ::toupper(words[i][0]);
str = Join(words, " ");

tamland commented:
It was more about having two levels of ternaries *with* assignments in them, but on second thought: passing non-pure function to `transform`, probably not a good idea. Transform will not necessary apply that function in-order.

And I think I completely misunderstood which kind of letter casing this was, sorry about that. It' the one listed as Title Case [here](http://en.wikipedia.org/wiki/Letter_case#Stylistic_or_specialised_usage) right? *Not* 'sentence case'.

ronie commented:
have you tested it with accented characters as well (since we had issues with those in the past)?

and now the bonus question, does it work corectly with RTL languages too?

mkortstiege commented:
@tamland, this one is Start case. Every first letter is capitalized regardless of function. Objections to change it to the Split/Join method?

@ronie, will test and report back.

Paxxi commented:
this doesn't work on Windows first because of the std:: I mentioned above and if that's changed to the c-versions it fails because it's not expecting utf8 strings and throws asserts. We also use the C locale everywhere and that is not suitable for these types of string conversions, it'll likely fail on anything except english.

You could make sure it's operating on wstrings and use the uppercase table that's available in stringutils or grab a locale according to user preferences and pass that to isspace and toupper to let it work with the correct casing rules.

mkortstiege commented:
Updated.

There's still an issues with uppercasing accented characters on FreeBSD, OSX (darwin) and uClibc-based platforms as they don't support `std::toupper` using `std::locale` due to [LangInfo.cpp#L175](https://github.com/xbmc/xbmc/blob/master/xbmc/LangInfo.cpp#L175).

@Fneufneu, @Memphiz mind checking whether or not this limitation is still valid?

Memphiz commented:
i delegate to @Karlson2k because i think he knows much more about this stuff

da-anda commented:
I agree with @tamland that capitalize is a bit misleading word for it - reading it at least I would expect this function to either make the string all caps (= uppercase) or small caps ("Kapitälchen" in German).  In PHP the function uppercasing every first letter of a word is called "ucword" (so uppercase word) - not the best either, but [UPPERASE_WORD] would be more selfexplaining. In same way we could add [UPPERCASE_FIRST] to make the first letter of a string caps.
I have no hard feelings about it and am also fine with keeping CAPITALIZE - only prefer blue for the bikeshed :)

tamland commented:
Your 'uppercase first' is 'sentence case', which is what I was taking about when I said that:P Actually, python uses 'capitalize' too so just pick one

Paxxi commented:
as @tamland said just pick a name that feels somewhat ok :)
I vote to get this merged as is, it doesn't work for some cases but it's not any worse than our current toupper methods and we can get it sorted out later on when/if we get around to fixing stringutils to handle real unicode.

I would like to see one or more test cases added though, at least one with spaces in the test string :)

@ronie it will work ok for RTL text that's in it's logical representation and represented as LTR with PDF markings to tell the rendererer to flip the string. It will not work if it's stored as RTL and uppercase the last character instead :)

mkortstiege commented:
Added an additional test string with spaces.

Karlson2k commented:
In text editor this style is usually named "Titlecase" as it commonly used in titles. I think that name is self-explainable.

Karlson2k commented:
For the code.
Firstly I'd like to stop producing functions for modifying text in-place. But currently this isn't inline with other transforming functions, so this is optional.
Secondly, locale is created for each letter.
I'd write this function simpler and straightforward:
```` c++
void StringUtils::ToTitlecase(std::wstring &str)
{
  const std::locale& loc = g_langInfo.GetLocale();
  bool isFirstLetter = true;
  for (std::wstring::iterator it = str.begin(); it < str.end(); ++it)
  {
    if (std::isspace(*it, loc))
      isFirstLetter = true;
    else if (isFirstLetter)
    {
      *it = std::toupper(*it, loc);
      isFirstLetter = false;
    }
  }
}
````

mkortstiege commented:
Thanks for you input. Will update this one tomorrow.

Karlson2k commented:
The rest looks fine. RTL doesn't matter because text stored in logical order and all RTL alphabets don't have uppercase letters.
Also keep in mind that some languages use combination of letters (digraphs and trigraphs), like 'ch' in Polish and Czech. Luckily for titlecase that doesn't matter. :)

mkortstiege commented:
Updated. Thanks again for the input.

mkortstiege commented:
jenkins build this please.

Karlson2k commented:
I missed test commit!
For complete testing I'd add tests with words with rAndOM case, several spaces before words, spaces at the start of the string, spaces at the end and add tabs and other delimiters.
After run-testing it's ready for merge. :)

topfs2 commented:
In css the term is correct at least, not sure how its done elsewhere.
http://www.w3schools.com/cssref/pr_text_text-transform.asp

topfs2 commented:
A few concerns, to upper isn't always good on all languages AFAIK, but we are using it for uppercase text already I guess so perhaps its out of scope for this PR.

Another one I think worth trying is sentences with double spaces, so double spacing is preserved. Also break lines should also trigger the casing I guess, but perhaps thats handles higher up in the text transform?

mkortstiege commented:
@topfs2, I think we don't have to worry about languages. It's the skinners decision to use it and to make it an optional skin setting if. We're just offering the ability.

Will check double spaces and line breaks while doing the unit tests.

tamland commented:
>In text editor this style is usually named "Titlecase" as it commonly used in titles. I think that name is self-explainable.

Sure, but that's not what this is though. Titlecase ignores articles etc. This and 'capitalize' does not.

mkortstiege commented:
I'll allow the paint to dry and continue this one whenever you guys are ready ;)

mkortstiege commented:
@Karlson2k added some weird tests. Not sure if it's that what you had in mind..

Karlson2k commented:
@tamland, "titlecase" means every word starts with Uppercase and rest is lowercase. "Capitalize" looks like all-uppercase. For me "titlecase" is selfexplainable, "capitalize" need additional explanation.
BTW, @mkorstiege if "uppercase" and "lowercase" converts all letters, don't we need to convert rest of the letters to lowercase to be inline with other functions?

mkortstiege commented:
@Karlson2k that was my first approach as well. Lower all and then upper first but as @HitcherUK pointed out on forums, it looks a bit meh for some like McDonalds -> Mcdonalds or NYPD becomes Nypd.

da-anda commented:
yes, it should only uppercase first letter and leave the rest unchanged. Btw - what's treated as word boundary? Only spaces so far, right? (ini case I read the code correctly). In any case we shouldn't uppercase string parts connected by a dash. So "Blu-ray" should stay as is and not become Blu-Ray. Maybe also add this as testcase?

Hope I didn't put too much paint on so that it's drying fast ;)

mkortstiege commented:
Blu-Ray won't happen (dash != space).

topfs2 commented:
Instead of trying to debate what you personally think is the best name, perhaps check what other tools name it. As that has a far larger likelihood of bring universal for more people. We don't always have to recreate the wheel :) as said in css its called capitalise. What is it in other GUI toolkits?

topfs2 commented:
Css also only transform first letter, I.e. leaving McDonalds and NYPD

tamland commented:
You can chain these functions right? Then only transforming first makes sense as it lets you achieve both.
@Karlson2k Just for reference what would you call the common style (in English at least) of Starting Major Words in Titles and Headers with Uppercase?

mkortstiege commented:
Squashed and rebased. Will change it to whatever you guys think is the right name.

For the record: I started with Capitalize and still think it's correct.

Montellese commented:
I vote for Capitalize because that's what CSS uses.

topfs2 commented:
Another concern is that in some languages you uppercase more than a single
letter http://en.m.wikipedia.org/wiki/Capitalization

This is something we would need language knowledge about, and that's not
always the easiest to know. Description tags which are scraped etc. Anyways
I thought it would be interesting to mention. IMO that is something we can
improve in a later PR and shouldn't block this PR.

Karlson2k commented:
Yep, use "Capitalize". Let's us it like in CCS.

mkortstiege committed:
[fonts] adds CAPITALIZE - capitalize the first letter of each word

mkortstiege commented:
Re-renamed to capitalize and rebased. jenkins build this please.

tamland commented:
Up next: capitalize or capitalise?:D

mkortstiege commented:
Z please but i have no strong opinion on this. Thanks for annoying the shit out of me ;)

da-anda commented:
lol. I personally don't care, but our default language is british English, so in case we really want do argue about z vs s I'd say use british spelling :)

HitcherUK commented:
> Up next: capitalize or capitalise?:D

The skinning engine seems to use American English (eg color, localize, etc) so I guess capitalize.

Karlson2k commented:
Let's name it exactly the same like in CSS.

mkortstiege commented:
jenkins build this please

topfs2 commented:
Android
http://developer.android.com/reference/android/widget/TextView.html#attr_android:capitalize

mkortstiege commented:
jenkins build this please

mkortstiege commented:
What´s the status here?

mkortstiege commented:
Will shove this one in tomorrow if there are no objections.

HitcherUK commented:
@mkortstiege the style tag for fonts this isn't working.

`<style>capitalize</style>`

mkortstiege commented:
Thanks, will check.

mkortstiege commented:
https://github.com/xbmc/xbmc/pull/6760 will take care of the style tag issue.

HitcherUK commented:
@mkortstiege Confirmed this works at the same time as the hint text, many thanks.

HitcherUK commented:
@mkortstiege Just tried to use this to capitalize a string (id=21429 / string=Order by) but instead of getting 'Order By' I just got '21429'. Tried to add LOCALIZE but that just gave me '[LOCALIZE]21429[/LOCALIZE]' instead.

Thanks.

mkortstiege commented:
Please post dummy code on the forums so i can easily reproduce. Thanks.

HitcherUK commented:
Will do.
